doni/driver/worker/blazar.py

```diff
@@ -1,20 +1,413 @@
+            if exc.code == 404:
+                # TODO Host id in lease doesn't exist, what do do?
+                return WorkerResult.Defer(result)
+        else:
+            return WorkerResult.Success(result)
+
     def process(
         self,
         context: "RequestContext",
         hardware: "Hardware",
         availability_windows: "list[AvailabilityWindow]" = None,
-        state_details: "dict" = None,
+        state_details: "dict" = {},
     ) -> "WorkerResult.Base":
-        pass
+        """Main loop for Blazar sync worker.
+
+        This method ensures that an up-to-date blazar host object exists for
+        each physical host in Doni's DB.
+
+        "name" must match the name used by nova to identify the node. In our case
+        it is the hardware uuid, as that is what ironic is passing to nova.
+        blazar uses nova.get_servers_per_host to check if there is an existing
+        server with that name.
+        """
+        # If we know the host_id, then update that host. Otherwise, attempt to create it.
+        host_id = state_details.get("blazar_host_id")
+        if host_id:
+            # TODO: We always update the host. We should add a precondition of some kind.
+            host_result = self._blazar_host_update(context, hardware, host_id)
+        else:
+            # Without a cached host_id, try to create a host. If the host exists,
+            # blazar will match the uuid, and the request will fail.
+            host_result = self._blazar_host_create(context, hardware)
+
+        if isinstance(host_result, WorkerResult.Defer):
+            return host_result  # Return early on defer case
+
+        # Get all leases from blazar
+        leases_to_check = self._blazar_lease_list(context)
+
+        lease_results = []
+        # Loop over all availability windows that Doni has for this hw item
+        for aw in availability_windows or []:
+            new_lease = _blazar_lease_requst_body(aw)
+            # Check to see if lease name already exists in blazar
+            matching_index, matching_lease = _search_leases_for_lease_id(
+                leases_to_check, new_lease
+            )
+            if matching_lease:
+                # Pop each existing lease from the list. Any remaining at the end will be removed.
+                leases_to_check.pop(matching_index)
+                # Update case.
+                if not new_lease.items() <= matching_lease.items():
+                    # If new lease is a subset of old_lease, we don't need to update
+                    lease_results.append(self._blazar_lease_update(context, new_lease))
+            else:
+                lease_results.append(self._blazar_lease_create(context, new_lease))
+
+        delete_results = []
+        # Delete any leases that are in blazar, but not in the desired availability window.
+        for lease in leases_to_check:
+            delete_results.append(self._blazar_lease_delete(context, lease))
```

Comment:
Whenever there is a list where a subset of its items (possibly all items) is being transformed and stored in a new list, a list comprehension is often the most elegant way to express it:

```suggestion
        delete_results = [self._blazar_lease_delete(context, lease) for lease in leases_to_check]
```

Reply:
(this doesn't really apply if using the approach highlighted in the other comment)